wire-library/wire-compiler/src/main/java/com/squareup/wire/WireCompiler.kt

```diff
@@ -143,66 +137,48 @@ class WireCompiler internal constructor(
       )
     }
 
-    Closer.create().use { closer ->
-      val sources = protoPaths.map { it.toPath() }
-      val directories = directoryPaths(closer, sources)
+    val sources = protoPaths.map { it.toPath() }
 
-      val allDirectories = directories.map { Location.get(it.key.toString()) }.toList()
-      val sourcePath: List<Location>
-      val protoPath: List<Location>
+    val allDirectories = sources.map { Location.get(it.toString()) }.toList()
+    val sourcePath: List<Location>
+    val protoPath: List<Location>
 
-      if (sourceFileNames.isNotEmpty()) {
-        sourcePath = sourceFileNames.map { locationOfProto(directories, it) }
-        protoPath = allDirectories
-      } else {
-        sourcePath = allDirectories
-        protoPath = listOf()
-      }
+    if (sourceFileNames.isNotEmpty()) {
+      sourcePath = sourceFileNames.map { locationOfProto(sources, it) }
+      protoPath = allDirectories
+    } else {
+      sourcePath = allDirectories
+      protoPath = listOf()
+    }
 
-      val wireRun = WireRun(
-          sourcePath = sourcePath,
-          protoPath = protoPath,
-          treeShakingRoots = treeShakingRoots,
-          treeShakingRubbish = treeShakingRubbish,
-          targets = targets,
-          modules = modules,
-          permitPackageCycles = permitPackageCycles
-      )
+    val wireRun = WireRun(
+        sourcePath = sourcePath,
+        protoPath = protoPath,
+        treeShakingRoots = treeShakingRoots,
+        treeShakingRubbish = treeShakingRubbish,
+        targets = targets,
+        modules = modules,
+        permitPackageCycles = permitPackageCycles
+    )
 
-      wireRun.execute(fs, log)
-    }
+    wireRun.execute(fs, log)
   }
 
-  /**
-   * Map the physical path to the file system root. For regular directories the key and the
-   * value are equal. For ZIP files the key is the path to the .zip, and the value is the root
-   * of the file system within it.
-   */
-  private fun directoryPaths(closer: Closer, sources: List<Path>): Map<Path, Path> {
-    val directories = mutableMapOf<Path, Path>()
-    for (source in sources) {
-      directories[source] = when {
-        Files.isRegularFile(source.toNioPath()) -> {
-          val sourceFs = FileSystems.newFileSystem(source.toNioPath(), javaClass.classLoader)
-          closer.register(sourceFs)
-          sourceFs.rootDirectories.single().toOkioPath()
-        }
-        else -> source
+  /** Searches [sources] trying to resolve [proto]. Returns the location if it is found. */
+  private fun locationOfProto(sources: List<Path>, proto: String): Location {
+    val directoryEntry = sources.find { source ->
+      when (fs.metadataOrNull(source)?.isRegularFile) {
+        true -> fs.openZip(source).exists("/".toPath() / proto)
```

Comment:
We should cache the results of `fs.openZip` because it’s quite expensive

Reply:
(That’s what directoryPaths did in the old code)

Reply:
I see. I think the old code was overly confusing because it cached stuff far away from here while at the same time, it's only used here.
I've inlined the caching inside this method. Let me know.